src/Databases/DatabasesCommon.cpp

```diff
@@ -31,8 +42,87 @@ namespace ErrorCodes
+    /// If there are no columns, then there is nothing much we can do
+    if (!new_query.columns_list || !new_query.columns_list->columns)
+        return;
+
+    const auto & columns = *new_query.columns_list;
+    /// Do some basic sanity checks. We cannot do the same strict checks as on create, because context might not have the same settings if it is not called directly from an alter query.
+    /// SECONDARY_CREATE should check most of the important things.
+    const auto columns_desc
+        = InterpreterCreateQuery::getColumnsDescription(*columns.columns, context, LoadingStrictnessLevel::SECONDARY_CREATE, false);
+
+    /// Default expressions are only validated in level CREATE, so let's check them now
+    DefaultExpressionsInfo default_expr_info{std::make_shared<ASTExpressionList>()};
+
+    for (const auto & ast : columns.columns->children)
+    {
+        const auto & col_decl = ast->as<ASTColumnDeclaration &>();
+        /// There might be some special columns for which `columns_desc.get` would throw, e.g. Nested column when flatten_nested is enabled.
+        /// At the time of writing I am not aware of anything else, but my knowledge is limited and new types might be added, so let's be safe.
+        if (!col_decl.default_expression)
+            continue;
+
+        /// If no column description for the name, let's skip the validation of default expressions, but let's log the fact that something went wrong
+        if (const auto * maybe_column_desc = columns_desc.tryGet(col_decl.name); maybe_column_desc)
+            getDefaultExpressionInfoInto(col_decl, maybe_column_desc->type, default_expr_info);
+        else
+            LOG_WARNING(getLogger("validateCreateQuery"), "Couldn't get column description for column {}", col_decl.name);
+    }
+
+    if (default_expr_info.expr_list)
+        validateColumnsDefaultsAndGetSampleBlock(default_expr_info.expr_list, columns_desc.getAll(), context);
 
-void applyMetadataChangesToCreateQuery(const ASTPtr & query, const StorageInMemoryMetadata & metadata)
+    if (columns.indices)
+    {
+        for (const auto & child : columns.indices->children)
+            IndexDescription::getIndexFromAST(child, columns_desc, context);
+    }
+    if (columns.constraints)
+    {
+        InterpreterCreateQuery::getConstraintsDescription(columns.constraints, columns_desc, context);
+    }
+    if (columns.projections)
+    {
+        for (const auto & child : columns.projections->children)
+            ProjectionDescription::getProjectionFromAST(child, columns_desc, context);
+    }
+    if (!new_query.storage)
+        return;
+    const auto & storage = *new_query.storage;
+
+    std::optional<KeyDescription> primary_key;
+    /// First get the key description from order by, so if there is no primary key we will use that
+    if (storage.order_by)
+        primary_key = KeyDescription::getKeyFromAST(storage.order_by->ptr(), columns_desc, context);
+    if (storage.primary_key)
```

Comment:
Should we check for `primary_key` here as well or not?

Reply:
I think no. Because if we have `storage.primary_key`, then we have to use that anyway. We also have to call `getKeyFromAST` for `storage.order_by` to ensure it is analyzed in the same way as on table creation.